### ISSUE TITLE
[Indiana Jones and the Infernal Machine] invisible save game list and dgVoodoo2 not working

### ISSUE THREAD

**Issue by Spatzendame** [narzoul/DDrawCompat#410]
Hello Narzoul!

I want to report two things in your ddraw wrapper that actually doesn't work with this game.
I tried so many different ddraw wrappers and yours is the only one working best with the window behavior of the game. For example when you are in game and open an option window or when you finish a level and the trade screen appears. All other ddraw wrappers lose focus and the game will become in trouble. Your file works best, but there are these two things that you should fix to make it perfect.
First is when you press "Resume game" and open the save game selection window, it won't show the save game files. The save game files are still there, but invisible. You have to click blindly in the window to select a save game. You can see it in the screenshot attachment.
Second problem is that your ddraw.dll will only recognize the standard primary Microsoft Direct3D Hardware acceleration HAL driver. That's not a problem, but I prefer to use dgVoodoo2 (with your ddraw.dll and not dgVoodoo2's ddraw.dll but with D3DImm.dll and D3D9.dll) and it will have no effect to the game. Maybe you can improve compatibility with dgVoodoo2?

For playing the game you need the GOG or the latest Steam version and a dinput.dll file from here:
https://github.com/riverar/IndirectInput

Maybe you can make an own, even better dinput.dll for this game? But that one works good.

Thank you, Narzoul. Maybe you can improve general compatibility with this fantastic game.

![indy3D_narzoul_ddraw](https://github.com/user-attachments/assets/dd5d0b8c-c952-47d8-9047-7ece297f0ebd)

**Comment by Spatzendame** [comment]
Oh, sorry. I haven't seen someone already reported about this issue.
Compatibility with dgVoodoo2 should be possible because GOG's own ddraw.dll which is part of the GOG version by default, works with it too. It's correct, that you can not use both ddraw.dll files at same time, so I only use D3DImm.dll and D3D9.dll from dgVoodoo2 (not the ddraw.dll because this has the same window issues as all other ddraw wrappers except yours). The thing is when I use dgVoodoo2 it has a lot of useful features that works when you have only D3D9.dll from dgVoodoo2 installed. There is no need to use any other file so it normally should work with a different ddraw.dll in combination.

But the main issue is the loading screen. When you can fix this, it would be fantastic. :-)

There is no problem with DirectInput. It's just this file was originally developed for another game and works only with coincidence with Indiana Jones and the Infernal Machine too. Just want to inform about that because without this dinput.dll even the GOG and Steam versions doesn't no longer start on Win11 24H2 for me. So you may need this file too when trying to fix ddraw.dll.

**Comment by BEENNath58** [comment]
> Second problem is that your ddraw.dll will only recognize the standard primary Microsoft Direct3D Hardware acceleration HAL driver. That's not a problem, but I prefer to use dgVoodoo2

This is the Direct3D Immediate Mode Hardware accelerated driver. dgVoodoo2 uses the same too, except it renames it and converts the output API to D3D10/11/12.

> Indiana Jones and the Infernal Machine

This game is from 1999. Since I currently don't have access to the GOG game, my assumption is that the GOG ddraw.dll is a DirectDraw -> Direct3D9 wrapper, since you mentioned: 
`"The thing is when I use dgVoodoo2 it has a lot of useful features that normally works when you have only D3D9.dll from dgVoodoo2 installed."`

The D3DImm.dll changes some of the 3D rendering parts, that DDrawCompat does in ddraw.dll.

In short, using ddraw.dll of DDC is equivalent to using ddraw.dll+d3dimm.dll from dgVoodoo2

What I didn't understand is why GOG ddraw.dll is interacting with d3dimm.dll, it shouldn't be since both are different proprietary libraries. It could be that deleting the d3dimm.dll will have no effect on the game whatsoever.

**Comment by narzoul** [comment]
> Compatibility with dgVoodoo2 should be possible because GOG's own ddraw.dll which is part of the GOG version by default, works with it too.

I was referring specifically to compatibility with DDrawCompat. I think GOG's wrapper maps to the Direct3D 9 runtime, which dgVoodoo can then wrap to the Direct3D 11/12 runtime by replacing d3d9.dll. But it's not the case with DDrawCompat, which hooks into Direct3D 9 at the user-mode display driver level, and otherwise runs on top of the native DirectX 1-7 runtimes, both of which are completely bypassed by dgVoodoo, so they'll never be compatibly (unless I rewrite DDrawCompat to also map to a DX9+ runtime, but I don't have thousands of hours to spend on that).

**Comment by Spatzendame** [comment]
Thank you both for explaining. Narzoul, how can I contact you privately?
Otherwise I will give you 5 dollars so you can buy the game? I would give you even a little donation in top of that for your investigation and fix the issues. I'm okay with the dgVoodoo part, but it would be nice if you could fix the issues with the game. Your DDrawCompat already works the best but the invisible loading screen is annoying.

**Comment by narzoul** [comment]
I looked into it, but the issue is not with DDrawCompat. That part of the save/load dialogs are not rendered even natively. It may be some bug in Windows 11, because I don't have the issue in a Windows 10 VM.

The game uses the GetOpenFileNameA/GetSaveFileNameA Windows API functions to open the load/save dialogs, but also adds some customizations using the OFN_ENABLEHOOK and OFN_ENABLETEMPLATE flags. Removing those flags fixes the problem, but it also removes the customizations. I think it's just the small thumbnail on the right right side of the dialogs that are lost. This is probably still better than having an invisible file list.

You can remove the problematic flags by patching Resource\Indy3D.exe with these two hex edits:
64 01 0A 00 -> 04 01 0A 00
65 01 0A 00 -> 05 01 0A 00
Both sequences should only appear once, at least in the GOG version.

This should make the open/save dialogs function as a standard open/save dialog, with working file list. It works with DDrawCompat and with GOG's wrapper, but crashes the game without wrappers for some reason when the dialogs are opened.

As this issue is unrelated to DDrawCompat and I don't want to add such non-generic hacks to it, I'll close this as "resolved".

**Comment by smlu** [comment]
Just for info, **OpenJones3D** (https://github.com/smlu/OpenJones3D/releases/tag/v0.3.0) fixes this overlay map rendering issue.
